src/Lean/Elab/Tactic/BVAckermannize.lean

```diff
@@ -0,0 +1,407 @@
+  Eq.trans (Eq.trans ackEqApp appEqApp) (Eq.symm ackEqApp')
+/-
+Make the ackermannization theorem, which states that: `(∀ i, arg₁[i] = arg₂[i]) -> call₁ = call₂`.
+Formally, we build an expr such as `arg₁ = arg'₁ -> arg₂ = arg'₂ -> ... argₙ = arg'ₙ -> call₁ = call₂`,
+where the proof is by congruence over the equalities.
+-/
+def mkAckThm (g : MVarId) (fn : Function) (args args' : Array Argument) (call call' : CallVal) : AckM MVarId := do
+    trace[bv_ack] "making ack congr thm for '{fn}' '{args}' ~ '{args'}',  calls '{call}', '{call'}'"
+    if args.size = 0 then
+      throwTacticEx `bv_ack g
+        m!"expected {args} to have more than zero arguments when building congr theorem for {fn}."
+    if args'.size = 0 then
+      throwTacticEx `bv_ack g
+        m!"expected {args'} to have more than zero arguments when building congr theorem for {fn}."
+
+    if args.size ≠ args'.size then
+      throwTacticEx `bv_ack g
+        m!"internal error: expected {args} to have the same size as {args'} when building congr thm for {fn}."
+
+    let mut eqHyps := #[]
+    for (arg, arg') in args.zip args' do
+      eqHyps := eqHyps.push (← mkEq arg.x arg'.x)
+    -- we build the equality according to the example above.
+    let mut localDecls : Array (Name × BinderInfo × (Array Expr → AckM Expr)) := #[]
+    let mut i := 0
+    for (arg, arg') in args.zip args' do
+      let name := Name.num (Name.mkSimple "ack_arg") i
+      localDecls := localDecls.push (name, BinderInfo.default, fun _ => mkEq arg.x arg'.x)
+    let ackEqn ← g.withContext <| withLocalDecls localDecls fun argsEq => do
+      let mut fEq ← mkEqRefl fn.f
+      for argEq in argsEq do
+        fEq ← mkCongr fEq argEq
+
+      let finalEq ← mkEqTrans (← mkEqTrans call.heqProof fEq) (← mkEqSymm call'.heqProof)
+      mkLambdaFVars argsEq  finalEq
+    trace[bv_ack] "made ackermann equation: {ackEqn}"
+    let (_ackEqn, g) ← g.note (Name.mkSimple s!"ackEqn{fn.f}") ackEqn
+    return g
+
+/-
+For every bitvector (x : BitVec w), for every function `(f : BitVec w → BitVec w')`,
+walk every function application (f x), and add a new fvar (fx : BitVec w').
+- Keep an equality that says `fx = f x`.
+- For function application of f, for each pair of bitvectors x, y,
+  add a hypothesis that says `x = y => fx = fy, with proof.
+-/
+def ack (g : MVarId) : AckM MVarId := do
+  g.withContext do
+    let (target', g) ← introAckForExpr g (← inferType (Expr.mvar g))
+    let g ← g.replaceTargetDefEq target'
+
+    let hyps ← g.getNondepPropHyps
+    let mut g := g
+    for hyp in hyps do
+      g ← g.withContext do
+        withTraceNode m!"🎯 hyp '{← hyp.getType}'" (collapsed := false) do
+          let hypG ← introAckForExpr g (← hyp.getType)
+          pure hypG.2
+
+    for (fn, arg2call) in (← get).fn2args2call do
+      let argCallsArr := arg2call.toArray
+      for i in [0:argCallsArr.size] do
+        let (arg₁, call₁) := argCallsArr[i]!
```

Comment:
If you are feeling fancy you can actually make these provably in bounds with the `for h : i in ...` syntax.

Reply:
Nice, thanks for the fanciness refactor